### ISSUE TITLE
Some problems about loss Convergence

### ISSUE THREAD

**Issue by Kite0011** [HazyResearch/flash-attention#114]
Thanks a lot for your great work!

I'm trying using flash attn to replace nv's megatron origin multi head attention, but while i'm trying to run it end to end ＆ observe its loss convergence, it differs from origin attention in afew steps later(about 1000)

I'm curious if you can show me the code how to add flash-attn into megatron, ＆ i'll put my implement at last.

I really appreciate your help, and I look forward to your reply as soon as possible!

<img width="507" alt="image" src="https://user-images.githubusercontent.com/50414751/215708906-b4d025d3-5e76-44b8-b648-39d8a03c1cab.png">

```
### in megatron/model/transformer.py file class ParallelAttention 
### here we step in need_unpad==True branch 
    @instrument_w_nvtx
    def forward(self, hidden_states, attention_mask, layer_past=None,
                get_key_value=False, encoder_output=None):
        # hidden_states: [sq, b, h]

        # =====================
        # Query, Key, and Value
        # =====================

        if self.attention_type == AttnType.self_attn:
            # Attention heads [sq, b, h] --> [sq, b, (np * 3 * hn)]
            mixed_x_layer, _ = self.query_key_value(hidden_states)

            # [sq, b, (np * 3 * hn)] --> [sq, b, np, 3 * hn]
            new_tensor_shape = mixed_x_layer.size()[:-1] + \
                (self.num_attention_heads_per_partition,
                 3 * self.hidden_size_per_attention_head)
            mixed_x_layer = mixed_x_layer.view(*new_tensor_shape)

            # [sq, b, np, 3 * hn] --> 3 [sq, b, np, hn]
            (query_layer,
             key_layer,
             value_layer) = mpu.split_tensor_along_last_dim(mixed_x_layer, 3)
        else:
            # Attention heads [sk, b, h] --> [sk, b, (np * 2 * hn)]
            mixed_kv_layer, _ = self.key_value(encoder_output)

            # [sk, b, (np * 2 * hn)] --> [sk, b, np, 2 * hn]
            new_tensor_shape = mixed_kv_layer.size()[:-1] + \
                (self.num_attention_heads_per_partition,
                 2 * self.hidden_size_per_attention_head)
            mixed_kv_layer = mixed_kv_layer.view(*new_tensor_shape)

            # [sk, b, np, 2 * hn] --> 2 [sk, b, np, hn]
            (key_layer,
             value_layer) = mpu.split_tensor_along_last_dim(mixed_kv_layer, 2)

            # Attention head [sq, b, h] --> [sq, b, hp]
            query_layer, _ = self.query(hidden_states)
            # [sq, b, hp] --> [sq, b, np, hn]
            new_tensor_shape = query_layer.size()[:-1] + \
                (self.num_attention_heads_per_partition,
                 self.hidden_size_per_attention_head)
            query_layer = query_layer.view(*new_tensor_shape)

        # ==================================
        # Adjust key and value for inference
        # ==================================

        if layer_past is not None:
            past_key, past_value = layer_past
            key_layer = torch.cat((past_key.type_as(key_layer),
                                   key_layer), dim=0)
            value_layer = torch.cat((past_value.type_as(value_layer),
                                     value_layer), dim=0)
        if get_key_value:
            present = (key_layer, value_layer)

        # ===================================
        # Raw attention scores. [b, np, s, s]
        # ===================================

        # [b, np, sq, sk]
        output_size = (query_layer.size(1),
                       query_layer.size(2),
                       query_layer.size(0),
                       key_layer.size(0))
        # origin attention in megatron-lm 
        if not self.use_flash_attention or not flash_attn_installed:
            # [sq, b, np, hn] -> [sq, b * np, hn]
            query_layer = query_layer.view(output_size[2],
                                        output_size[0] * output_size[1], -1)
            # [sk, b, np, hn] -> [sk, b * np, hn]
            key_layer = key_layer.view(output_size[3],
                                    output_size[0] * output_size[1], -1)

            # preallocting result tensor: [b * np, sq, sk]
            matmul_result = torch.empty(
                output_size[0]*output_size[1],
                output_size[2],
                output_size[3],
                dtype=query_layer.dtype,
                device=torch.cuda.current_device())

            # Raw attention scores. [b * np, sq, sk]
            matmul_result = torch.baddbmm(
                matmul_result,
                query_layer.transpose(0, 1),   # [b * np, sq, hn]
                key_layer.transpose(0, 1).transpose(1, 2),  # [b * np, hn, sk]
                beta=0.0, alpha=(1.0/self.norm_factor))

            # change view to [b, np, sq, sk]
            attention_scores = matmul_result.view(*output_size)

            # ==================================================
            # Update attention mask for inference. [b, np, sq, sk]
            # ==================================================

            if get_key_value:
                with torch.no_grad():
                    if layer_past is not None:
                        attention_mask = attention_mask[
                            ...,
                            attention_scores.size(3) - 1,
                            :attention_scores.size(3)].unsqueeze(2)
                    else:
                        attention_mask = attention_mask[
                            ...,
                            :attention_scores.size(3),
                            :attention_scores.size(3)]

            # ===========================
            # Attention probs and dropout
            # ===========================

            # attention scores and attention mask [b, np, sq, sk]
            attention_probs = self.scale_mask_softmax(attention_scores,
                                                    attention_mask)

            # This is actually dropping out entire tokens to attend to, which might
            # seem a bit unusual, but is taken from the original Transformer paper.
            with mpu.get_cuda_rng_tracker().fork():
                attention_probs = self.attention_dropout(attention_probs)

            # =========================
            # Context layer. [sq, b, hp]
            # =========================

            # value_layer -> context layer.
            # [sk, b, np, hn] --> [b, np, sq, hn]

            # context layer shape: [b, np, sq, hn]
            output_size = (value_layer.size(1),
                        value_layer.size(2),
                        query_layer.size(0),
                        value_layer.size(3))

            # change view [sk, b * np, hn]
            value_layer = value_layer.view(value_layer.size(0),
                                        output_size[0] * output_size[1], -1)

            # change view [b * np, sq, sk]
            attention_probs = attention_probs.view(output_size[0] * output_size[1],
                                                output_size[2], -1)

            # matmul: [b * np, sq, hn]
            context_layer = torch.bmm(attention_probs, value_layer.transpose(0, 1))

            # change view [b, np, sq, hn]
            context_layer = context_layer.view(*output_size)

            # [b, np, sq, hn] --> [sq, b, np, hn]
            context_layer = context_layer.permute(2, 0, 1, 3).contiguous()

            # [sq, b, np, hn] --> [sq, b, hp]
            new_context_layer_shape = context_layer.size()[:-2] + \
                (self.hidden_size_per_partition,)
            context_layer = context_layer.view(*new_context_layer_shape)
        # use flash_attn replace origin attention
        else:
            """
            now we have three tensor:
            query_layer: [sq, b, np, hn]
            key_layer: [sk, b, np, hn]
            value_layer: [sk, b, np, hn]
            and attention mask:
            attention_mask: [b, np, sq, sk] (padding) or [1, 1, sq, sk] (causal)
            """
            if self.device == None:
                self.device = query_layer.device
            q_size = query_layer.size()
            batch_size = q_size[1]
            seqlen = q_size[0]
            head_num = q_size[2]
            kv_channel = q_size[3]
            k_seqlen = key_layer.size(0)

            # extract attention mask q and attention mask k from attention_mask
            # todo(kitefang): support seq-pack in flash-attention
            # mask type 1: padding
            if self.need_unpad:
                if not self.casual:
                    #attention_mask_q = (attention_mask[:, 0, :, 0] == 1).repeat(batch_size, 1)
                    #attention_mask_k = (attention_mask[:, 0, 0, :] == 1).repeat(batch_size, 1)
                    attention_mask_q = (attention_mask[:, 0, :, 0] == 1)
                    attention_mask_k = (attention_mask[:, 0, 0, :] == 1)
                # mask tyoe 2: causal
                # attention_mask here is only [1, 1, sq, sk] when it's type is causal, and it value is like an inverted triangle
                # but we need trans attention_mask_q shape to [bs, seq]
                # so here we set some fake mask supposing no padding in gpt-like training.
                else:
                    attention_mask_q = torch.ones(batch_size, self.q_seqlen, device=self.device)
                    attention_mask_k = torch.ones(batch_size, self.k_seqlen, device=self.device)
                # unpad q k v
                # s: seq len
                # b: batch size
                # h: head num
                # d: kv channel dim
                # rearrange(query_layer, 's b h d -> b s h d')
                # rearrange(key_layer, 's b h d -> b s h d')
                # rearrange(value_layer, 's b h d -> b s h d')
                # qkvs: [s b h d -> (b s) h d]
                # cu_seqlens looks like: tensor([   0,  512, 1024, 1536, 2048, 2560, 3072, 3584, 4096], device='cuda:0',dtype=torch.int32)
                # while len(cu_seqlens) equals to (batch_size+1), each element equals to previous one plus seq_len
                #import pdb
                #pdb.set_trace()
                query_layer_unpad, query_indices, q_cu_seqlens, q_max_s = unpad_input(rearrange(query_layer, 's b h d -> b s h d'), attention_mask_q)
                key_layer_unpad, key_indices, k_cu_seqlens, k_max_s = unpad_input(rearrange(key_layer, 's b h d -> b s h d'), attention_mask_k)
                value_layer_unpad, value_indices, v_cu_seqlens, v_max_s = unpad_input(rearrange(value_layer, 's b h d -> b s h d'), attention_mask_k)
            else:
                query_layer_unpad = query_layer.transpose(0, 1).contiguous().view(batch_size*seqlen, head_num, kv_channel)
                key_layer_unpad = key_layer.transpose(0, 1).contiguous().view(batch_size*k_seqlen, head_num, kv_channel)
                value_layer_unpad = value_layer.transpose(0, 1).contiguous().view(batch_size*k_seqlen, head_num, kv_channel)
                q_cu_seqlens = torch.arange(0, (batch_size + 1) * seqlen, step=seqlen, dtype=torch.int32, device=self.device)
                k_cu_seqlens = torch.arange(0, (batch_size + 1) * k_seqlen, step=k_seqlen, dtype=torch.int32, device=self.device)
                q_max_s = seqlen
                k_max_s = k_seqlen

            # output: (b s) h d
            output_unpad = flash_attn_unpadded_func(query_layer_unpad, key_layer_unpad, value_layer_unpad,
                                              q_cu_seqlens, k_cu_seqlens, q_max_s, k_max_s,
                                              dropout_p=self.attention_dropout_prob,
                                              softmax_scale=self.softmax_scale, causal=self.casual,
                                              return_attn_probs=False)

            if self.need_unpad:
                # pad output back
                # (b s) h d
                output = pad_input(output_unpad, query_indices, batch_size, seqlen)
            else:
                output = output_unpad.view(batch_size, seqlen, head_num, kv_channel)

            context_layer = rearrange(output, 'b s h d -> s b (h d)', s=seqlen)

        # =================
        # Output. [sq, b, h]
        # =================

        output, bias = self.dense(context_layer)

        if get_key_value:
            output = [output, present]

        return output, bias
```

**Comment by tridao** [comment]
Thanks for trying it out.
We've trained BERT with FlashAttention we where match the convergence of the standard implementation (e.g, MLPerf benchmark 2.0 and 2.1). 
We have an optimized BERT implementation [here](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/models/bert.py) where we [test](https://github.com/HazyResearch/flash-attention/blob/06da275bcba0f4554f205f52fa20f74309f2d973/tests/models/test_bert.py#L89) it matches the output of the Hugging face implementation.
We also have [integrated](https://github.com/NVIDIA/Megatron-LM/blob/285068c8108e0e8e6538f54fe27c3ee86c5217a2/megatron/model/transformer.py#L318) FlashAttention in Megatron-LM (for causal attention, e.g., GPT training, for now).

Here are some suggestions:
- When you call `flash_attn_unpadded_func`, instead of `dropout_p=self.attention_dropout_prob`, set `dropout_p=self.attention_dropout_prob if self.training else 0.0` so that dropout is not applied during validation.
- Can you check that `self.softmax_scale` is `1 / math.sqrt(head_dim)`?
- Can you print out the result `context_layer` with/without FlashAttention? The two should be approximately the same, up to numerical precision. What we usually compare is:
(FlashAttention in fp16/bf16 - standard impl in fp32) vs (standard impl in fp16/bf16 - standard impl in fp32).
The two numerical errors are usually around the same.

**Comment by Kite0011** [comment]
Thanks for your reply!

1. `attention_dropout_prob` is quite an important  point and i'll follow your suggestion! But it might not the solution for current problem
2. `self.softmax_scale` here is :
```
self.norm_factor = math.sqrt(self.hidden_size_per_attention_head)
self.softmax_scale = self.norm_factor
```
i guess i've found what problem is. i'll try modify the scale and restart the experiment.
3. i've write an unit test for flash_attn and i found its relative difference between bf16 flashattn and bf16 originattn is very huge, i'll put result at last.
i can understand your formula means, but the difference shown below makes me feel like flash-attn is another kind of mha.
<img width="881" alt="企业微信截图_2291b6d3-85aa-49f6-a4f5-39c1089e6108" src="https://user-images.githubusercontent.com/50414751/215942746-4a2156ed-8fac-4b4a-9db2-43a47a89f776.png">

Last of the last, thanks a lot for your help!

**Comment by tridao** [comment]
Thanks! 
Can you get 
(FlashAttention in fp16/bf16 - standard impl in fp32) vs (standard impl in fp16/bf16 - standard impl in fp32)?

Keep in mind bfloat16 generally has large numerical errors (compared to fp16 or fp32).
As an example, if `x` is your output using standard attention in bf16, you can print out:
`x` vs `x + 0.3 - 0.3`.
You'll see that any floating point arithmetic on bf16 will have pretty large error.

**Comment by Kite0011** [comment]
Thanks for your Explanation! And in another task (text to img) we just found flash attn has the same loss convergence with origin attn.

But back to bert task, after modify the softmax_scale, its loss convergence still differs from origin one and doesn't converge.

When i compare my implement with yours, i find that you gpt seems not using func `flash_attn.bert_padding.unpad_input` ＆ `flash_attn.bert_padding.pad_input`. I'm not sure if this will cause any problems on loss convergence because gpt may still pad its input sequence to a fixed length?

i'll try your implement on gpt and see if its loss can reduce normally.
(our gpt model face the same loss problem as bert model face)

**Comment by Kite0011** [comment]
By the way, why your gpt implement on megatron says it doesn't support mask==padding?

**Comment by tridao** [comment]
I've found it easier to debug by printing out the output difference at each step.

Our GPT implementation assumes there's no padding.

For BERT, if there's key_padding_mask, for max efficiency, our BERT implementation [removes](https://github.com/HazyResearch/flash-attention/blob/06da275bcba0f4554f205f52fa20f74309f2d973/flash_attn/models/bert.py#L138) all the padding tokens ("unpad") before the very first layer, and only add them back after the last layer. This way there's no wasted computation on the padding tokens.

**Comment by Kite0011** [comment]
It‘s indeed more convenient to debug step by step. But due to the error caused by half precision and different computer between attns, I can hardly figure out who output the difference.

And I've tried your implement yesterday, it seems work well on gpt training! I will do a comparative experiment today.

I will synchronize info with you if there is any progress in the future!

After all, thanks and hope everything goes well with your work!

**Comment by Kite0011** [comment]
Your implement works well on gpt model, i'll check my implement later.

I still have a problem attention mask:
+ Is there a confilct of principle when i want to send an attention mask into flash-attn. I wanna to modify flash-attn kernel to support sending mask into attn computer, cause we have some strangely shaped masks which can not regard as pad or causal.

Here is the loss comparation on gpt.
<img width="615" alt="image" src="https://user-images.githubusercontent.com/50414751/216542112-bfb64988-7775-4dfe-bda0-da27c053f1d8.png">

Thanks
